src/Build/Graph/ProjectGraph.cs

```diff
@@ -60,14 +65,103 @@ public ProjectGraph(string entryProjectFile)
+            while (edgesToVisit.Count > 0)
+            {
+                ProjectGraphNodeWithTargets nodeWithTargets = edgesToVisit.Dequeue();
+                ProjectGraphNode node = nodeWithTargets.Node;
+                List<string> targets = nodeWithTargets.Targets;
+
+                targetLists[node].AddRange(targets);
+
+                // No need to continue if this node has no project references.
+                if (node.ProjectReferences.Count == 0)
+                {
+                    continue;
+                }
+
+                // Based on the entry points of this project, determine which targets to propagate down to project references.
+                var targetsToAppendToChildren = new List<string>();
+                ICollection<ProjectItemInstance> projectReferenceTargets = node.Project.GetItems(ProjectReferenceTargetsItemType);
+                foreach (var target in targets)
+                {
+                    foreach (var projectReferenceTarget in projectReferenceTargets)
+                    {
+                        if (projectReferenceTarget.EvaluatedInclude.Equals(target, StringComparison.OrdinalIgnoreCase))
+                        {
+                            string targetsMetadataValue = projectReferenceTarget.GetMetadataValue(ProjectReferenceTargetsMetadataName);
+                            targetsToAppendToChildren.AddRange(ExpressionShredder.SplitSemiColonSeparatedList(targetsMetadataValue));
+                        }
+                    }
+                }
+
+                // Queue the project references for visitation, if the edge hasn't already been traversed.
+                foreach (var projectReference in node.ProjectReferences)
+                {
+                    var projectReferenceEdge = new ProjectGraphNodeWithTargets(
+                        projectReference,
+                        HydrateDefaultTargets(projectReference.Project, targetsToAppendToChildren));
+                    if (visitedEdges.Add(projectReferenceEdge))
+                    {
+                        edgesToVisit.Enqueue(projectReferenceEdge);
+                    }
+                }
+            }
+
+            // Dedupe target lists.
+            foreach (KeyValuePair<ProjectGraphNode, List<string>> pair in targetLists)
+            {
+                List<string> targetList = pair.Value;
+                HashSet<string> seenTargets = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                int i = 0;
+                while (i < targetList.Count)
+                {
+                    if (seenTargets.Add(targetList[i]))
+                    {
+                        i++;
+                    }
+                    else
+                    {
+                        targetList.RemoveAt(i);
+                    }
+                }
+            }
+
+            // Massage the final dictionary into string arrays instead of lists.
```

Comment:
Yea, this kind of sucks.

`BuildRequestData` only takes an array for the target list. What's more is that it defensively makes a copy of it since arrays are mutable and it could be user input when used programatically.

Furthermore, we can't add an internal ctor for this case since `XMake` is in a difference assembly than `BuildRequestData`.

We *could* add a new overload (well, a bunch of overloads) which take an immutable list since that guarantees that it can't be mutated (read only collection isn't enough; the concrete is still mutable so the caller might still do that). But that would be a significant API change.

So in the end I just decided to do this :(

Reply:
If we made this return an `ImmutableList<string>` instead, then the caller would need to do the array conversion anyway. However, maybe that's OK since at least it pushes the allocation out rather than always doing it. It's also just better to have immutability in a public API.

Potentially as we progress this feature a bit more, it may not even end up being public anyway, which means we could potentially further optimize then.

Will change the API.